### ISSUE TITLE
self-modifying code isn't executed properly

### ISSUE THREAD

**Issue by kajott** [MiSTer-devel/ao486_MiSTer#33]
For example, consider the following code snippet:

~~~asm
  inc byte ptr setvalue+1
setvalue:
  mov al, 0
~~~

The first instruction is supposed to change the second one from "mov al, 0" to "mov al, 1", but ao486 still executes the original "mov al, 0" instruction. The modified instruction is only executed if there are a lot of other instructions inbetween (I need at least 41 `nop`s in my test here, but this might be dependent on the alignment of the code in memory).

Normally, I wouldn't care much, but unfortunately, this affects some real-world software. For example, **every program that's compiled with Turbo Pascal 7.0 crashes when using the `Intr` function** of the standard `DOS` unit. This is because that function can be used to call an arbitrary interrupt, but x86 doesn't have a `int r/m8` instruction, just `int imm8`. So the runtime code contains an `int 0` instruction that's patched with the specific interrupt number that is going to be called. However, ao486 still executes the `int 0`, which is the "division by zero" interrupt, and hence the program terminates due to a division by zero that never happened.

Note that `int` call patching _does_ work if EMM386's EMS emulation is used, so paging or protection help in that specific case; however, the simple `mov` instruction patching example above behaves identically regardless of which mode the CPU is in.

**Comment by jsmolina** [comment]
Does it have any relation with interrupt table complains on dos32/a?
https://github.com/MiSTer-devel/ao486_MiSTer/issues/32

**Comment by kajott** [comment]
No idea. I doubt that there's a relation between this and #32, but I wouldn't rule it out either.

**Comment by MobyGamer** [comment]
In case there's any doubt that this code is an edge case only applicable to early x86 CPUs, the program where this was discovered records the system and CPU where it has been run, which shows success on all Intel x86 CPUs and third-party clones (AMD K-series, cached upgrade boards, overdrive processors, even systems that have no business running DOS programs such as a Core i7-920).

**Comment by RobertPeip** [comment]
This is no wonder, as the prefetch pipeline of the CPU doesn't care for instructions to be modified. So if it was already prefetched (and the next ~5 instructions are ALWAYS prefetched) it uses the unmodified code.

This should have been always in the code, but would be good if you could verify with the previous release.

That being said: there is a possibility to fix this.
Easiest would be to offer an OSD option for prefetch off, where every instruction will flush the pipeline. Speed should be around 10-20% of current max speed.
If that is to slow, we would need to auto detect writes to the current codepage. This is easier in real mode, but in protected mode could be difficult.

In any case: please try your test with the previous release, just to rule that out.

**Comment by kajott** [comment]
I cooked up a (very inefficient) [small test program](https://github.com/MiSTer-devel/ao486_MiSTer/files/5098920/smctest.zip) that consists of many instances of
~~~asm
    nop  ; repeated a variable number of times to align and shift phase
    mov byte ptr .i + 1, GOOD_VALUE
    nop  ; repeated a variable number of times
.i: mov ah, BAD_VALUE
~~~
with 0 to 63 NOPs, and with the whole snippet aligned to 16 bytes (one cacheline) plus an additional 0-15 NOPs to test various "alignment phases". I also ran the current TOPBENCH.EXE to check whether that crashes. Here are the results:
- `ao486_20200510.rbf`: 19-48 NOPs required. Some variation across runs. TOPBENCH crashes.
- `ao486_20200816.rbf`, L1$ on, L2$ on: 16-43 NOPs required. Little variation. TOPBENCH crashes.
- `ao486_20200816.rbf`, L1$ on, L2$ off: 16-39 NOPs required. Regular pattern, no variation. TOPBENCH crashes.
- `ao486_20200816.rbf`, L1$ off: 3-6 NOPs required. Trivial pattern, no variation. TOPBENCH runs. No difference whether L2$ is on or off.

So, the issue is indeed older and hasn't been introduced with the current release. Disabling L1$ helps, but costs 30%-70% performance, depending on the benchmark.

The goal shouldn't be to get this down to zero (i.e. "self-modifying code just works, no matter what") at the expense of performance. Real x86 CPUs also need some distance between the code that performs modifications and the code that is being modified; even the 8086 had a prefetch buffer of a few bytes. I don't know how large that "minimum SMC distance" is on real 486s (and I have none to test on), but it's certainly lower than ao486's, if possibly only by a tiny amount. Anyway, it's enough to kill Turbo Pascal 7.0's run-time library ...

**Comment by RobertPeip** [comment]
Thanks for the research!

That matches with my thoughts of how it should be: prefetch buffer is up to 15 Dwords = 60 Bytes + processing pipeline containing few instructions. As your nop most likely will be converted to a byte nop, which is executed in 3 cycles, it all makes sense.

If some instructions between are allowed to be "old", then it may be enough to just invalidate the prefetch buffer upon a snoop.
This would behave about the same as L1 cache off, maybe a little (4-8 bytes/NOPs) more.
Do you think this could work?

**Comment by kajott** [comment]
That would be the ideal solution I think (and I can well imagine that this is also how it worked in a "real" ASIC implementation).

The Turbo Pascal library code where we noticed the issue has exactly 32 bytes of other instructions between the `mov` and the `int` it modifies (which might be deliberate, but it looks like it's pure coincidence; they didn't arrange instructions in unusual ways to get to that exact distance). This means that if we could only shave 11 bytes off (from 43 to 32), we'd be fine.

**Comment by maikmerten** [comment]
@kajott For what it's worth, this is the output of your test program on a VIA-C7, which is a 486esque-but-on-steroids scalar, single-issue (IIRC), in-order architecture:

![via-c7](https://user-images.githubusercontent.com/2611432/91150550-d16c5780-e6bc-11ea-9522-236c2e728eaa.jpg)

**Comment by kajott** [comment]
@maikmerten Cool, thanks! "All green" means that no NOPs are required to make the modified instruction become valid. Does anybody have, by chance, a real 486 to test on?

**Comment by maikmerten** [comment]
@kajott I do have a mighty 486DX in my office at work (public administration*). I copied over your test program to a floppy and can carry it there.

(*) well, it's a demonstration object, not a work machine.

**Comment by maikmerten** [comment]
@kajott This is the result for my Intel 486 DX/33:

![486-dx33](https://user-images.githubusercontent.com/2611432/91176076-0d65e380-e6e2-11ea-981b-5b92648c7434.jpg)

These results are stable over several runs.

It appears that starting with the Pentium (and AMD K5 etc.) the prefetch queue is kept consistent with regards to self-modifying code more thoroughly: http://computer-programming-forum.com/46-asm/41735e0aca24cb94.htm

This "happens" to coincide with the arrival of L1 instruction caches...

**Comment by wishusknight** [comment]
It would be interesting to see what that little program does on a Cyrix DLC/SLC. And for giggles porting it to 68K where the 020 has a singular tiny L1 Icache with a short prefetch queue of 1 instruction direct from l1 with no buffer, and I *think the 030 has separate but tiny L&D caches with a much deeper prefetch buffer separate from the Icache IIRC.... But I am absolutely not sure on the 030.

**Comment by RobertPeip** [comment]
I finally found time to look at this and implemented a prefetch Fifo reset on successful L1 cache snoop.
This is currently not optimal, as it always clears the prefetch Fifo on snoop, but the speed can be improved, if it is only cleared when the address written is currently inside this Fifo. But for now, you should be able to test if that solves the problem.

Below is the status before and after and rbf to test.  Unfortunatly, there are some situations where the Fifo clears seems to come delayed. However, the worst case situation is now better than on the Intel 486 DX/33, which had 30 NOPs in worst case and we now have 27, before it was 52. Hopefully this is good enough.

[ao486_selfmodify1.zip](https://github.com/MiSTer-devel/ao486_MiSTer/files/5145003/ao486_selfmodify1.zip)

![DSC_2004](https://user-images.githubusercontent.com/53250236/91633676-a2baed80-e9ea-11ea-9274-a51bff324c9e.JPG)
![DSC_2005](https://user-images.githubusercontent.com/53250236/91633678-a51d4780-e9ea-11ea-8e5e-32a7e472cc6b.JPG)

**Comment by kajott** [comment]
@RobertPeip Great job, the new version works perfectly here! In fact, it's even a very tiny bit _faster_ than the last release in multiple benchmarks ... did you change anything else?

-------

In other news, I updated the SMC test application:
**Download:** [smctest2.zip](https://github.com/MiSTer-devel/ao486_MiSTer/files/5146052/smctest2.zip)
**Changelog:**
- The test now runs continuously, updating as fast as possible. This is useful to see how stable the results are across thousands of runs.
- Alignment is now _actually_ done with NOPs instead of whatever weird shit YASM put there (`lea si,[si]`, jumps to the end of the alignment block etc.). This changes the results a bit (for the worse), as those occasional unsolicited jumps cleared the prefetch queue.
- More verbose screen output, including readings of the minimum and maximum SMC distance.

(Fun fact: Doing a ton of `align 16, nop` is no problem with YASM. Doing a ton of `align 16, db 90h` however makes the assembler _incredibly_ slow: from 100 ms to 25 seconds ...)

**Comment by RobertPeip** [comment]
I cannot see any speed gain in the benchmarks i used, but also no slow down.
(pc player, superscape, landmark, system information)

Do you have any chance to modify your program, so that a failed code change with value > x will result in some recognizable action?

Something like: writing the failed code distance to a register, then clear it immidiatly.
E.g. Distance 30 has failed -> set EDX to 30, then set it back to 0.
This would need to be as close to the failed check as possible.

This way i could trigger in Signaltap on such failed situations and see what the reason for those is, hoew they differ from the normal case.
I cannot promiss anything, but with this help, I could maybe bring it down to reproducible 14bytes.

**Comment by kajott** [comment]
@RobertPeip OK, I tried something. 
[smcdbg.com](https://github.com/MiSTer-devel/ao486_MiSTer/files/5146872/smcdbg.zip) is basically the same thing as `smctest`, except that it doesn't do anything that's observable on the outside. The only thing it does is modify the DX register: It is kept at zero all the time, but if a modification fails, it receives the failing SMC distance for the duration of one instruction. Effectively, it's an endless run of
~~~asm
    nop  ; lots of them
    mov byte ptr [.i + 1], 0  ; 0 = SMC OK
    nop  ; some of them
.i: mov dl, number_of_nops_above
    xor dl, dl  ; reset DL immediately
~~~
Care has been taken to minimize any effects that could modify DX: Interrupts are off, and the keyboard is polled using port 60h. If everything worked out (I can't check that much from here), DX should be well-behaved as long at `smcdbg.com` runs.

**Comment by RobertPeip** [comment]
Thanks a lot, i will try that out and report back the next days.

**Comment by RobertPeip** [comment]
Your program worked great!
With it, I could reduce it to 13 bytes max.

Would be great if you could also do some more tests before i commit that.

[ao486_prefetchfix2.zip](https://github.com/MiSTer-devel/ao486_MiSTer/files/5151104/ao486_prefetchfix2.zip)

![20200831_112701-screen](https://user-images.githubusercontent.com/53250236/91735956-5075fa00-ebad-11ea-9739-8483efcfa37c.png)

**Comment by maikmerten** [comment]
Just for comparison, here's the 486 DX33 again:

![smctest2-486dx33](https://user-images.githubusercontent.com/2611432/91748361-28dc5d00-ebc0-11ea-88b6-9634a0026f0b.jpg)

@RobertPeip great work!

**Comment by sorgelig** [comment]
As i guess the best case is when everything is green. We shouldn't look at real i486 behavior. Our target is a whole range from 8086 to 486 (besides the performance) if we want a maximum compatibility.

**Comment by MobyGamer** [comment]
If the goal of the ao486 core is broad x86 compatibility rather than 486 correctness, then the README.md should probably be modified to state this up front.  Until sorgelig's statement, I was under the impression that true i486dx compatibility was the goal.

**Comment by maikmerten** [comment]
As far as I understand, the target is not to mimic the i486, but to ensure that the SMC distance is reliably "small enough" to not matter. The i486 is certainly "compatible enough" with regards to self-modifying code. If the SMC distance of ao486 is not exceeding that observed on the i486, then things should be "at least as fine as on i486", which means pretty much "perfect".

I'm sure that not even the 8086 and 8088 are "completely green". In fact the SMC distance can be used  to tell 8086 and 8088 apart, due to the shorter prefetch queue on the latter:

> Differentiating between 8088s and 8086s is trickier. The easiest way I've found to do it is to modify code that's five bytes ahead of IP. Since the prefetch queue of an 8088 is four bytes and the prefetch queue of an 8086 is six bytes, an instruction five bytes ahead of IP won't have any effect on an 8086 the first time around.

[https://www.drdobbs.com/embedded-systems/processor-detection-schemes/184409011](https://www.drdobbs.com/embedded-systems/processor-detection-schemes/184409011)


Only starting with the Pentium the prefetch queue will be flushed on self-modifying code:

> The prefetch queue remains in use by Intel although it is not much discussed. The Intel 80386 early versions had a 16 byte queue that was reduced to 12 bytes in later steppings. Pentiums seem to have prefetch queues -- sometimes one for each pipeline. Where documented, they seem to be 32 bytes. It is claimed that Pentiums and clones thereof will detect self modifying code and void the prefetch queue to prevent using an old value. That precludes software detection of the prefetch queue by traditional methods.

[http://donaldkenney.x10.mx/GLOSSARY/PREFETCH.HTM](http://donaldkenney.x10.mx/GLOSSARY/PREFETCH.HTM)

(apparently Intel needed to cut down on the i386 prefetch queue to make sure the increased SMD distance didn't result in problems with preexisting real-life code)

Long story short: Reaching for "all green" is only necessary if Pentium-or-later behavior is needed.

**Comment by kajott** [comment]
@RobertPeip: The new bitstream works like a charm! No regressions so far, neither in functionality nor in performance.
(By the way, how did you make that screenshot?)

@sorgelig: The thing is, "maximum compatibility" isn't reached when everything is green. Maximum compatibility is reached when MiSTer behaves *exactly* like the system it is supposed to emulate, with all minute details -- stuff like the exact depth of the 8088 and 8086 prefetch queues, like the 286 LOADALL instruction, like the exact timing of the 486 pipeline. As far as I understand, that level of accuracy was never the goal of ao486 (unlike e.g. Minimig, which _does_ try to replicate its target system as closely as possible). All it should do is run software of its era reasonably well -- and software of the 8088-486 era had to live with a few extra instruction slots for self-modifying code anyway, as @maikmerten's results from an actual 486DX show.

**Comment by sorgelig** [comment]
I appreciate that you explain me what is the goal of ao486. I didn't know it.

**Comment by RobertPeip** [comment]
Not sure if there is software, that would fail when instruction right after the current instruction pointer are already modified, but it's possible. I know some examples from the GBA, where this is the case.

However, with PC, all those instruction timing and little details usually don't matter much, as there where far too many different behaving hardware.

Having the core behave exactly like a 486 would probably result in additional cores for older platforms and Pentium like platforms.
Not that this is impossible, but it's lot's of additional work, not to mind the work of getting a cycle accurate 486 in the first place.
If someone is going to do that, that's good.
Just the commitment of other developers than sorgelig in the last years was very very low in the last years, so this is not something to rely on.

As for this bug here: we reached a good state now. It's reasonable low, like best case of the 486 DX33. If it also introduces no new bugs, we can probably keep it for the time being.

**Comment by dwillmore** [comment]
Greetings. I came here by way of a blog post by Ken Sherriff where KeyJ said you needed runs from actual Pentiums.  I seem to have a wide selection of olde processors lying around most of which I should be able to test--I think I have MBs for them.  Since they mentioned Pentiums, I took a look and found A8050166, A8050290, A80502133, A80502150, and FV80503233.  I also have similar processors from the 486 and 386 generations from Intel, IBM, Cyrix, AMD, and WinChip.  I can look up their models as well if it would be of use.  Is there a FreeDOS floppy image with the test program available?  I think I could find a way to get a modern machine to write to a floppy, bit it would take more work than to get these chips up and running. :)

**Comment by RobertPeip** [comment]
I haven't heard of any complaint about the current status of self-modifying code on the AO486.
Instead i wonder why this bug is still open :)

Feel free to do the research and comparison, but from my side i will most likely not try to optimize this further.

**Comment by kajott** [comment]
Indeed I think there's a slight misunderstanding here. @dwillmore is referring to [a comment](http://www.righto.com/2023/01/inside-8086-processors-instruction.html?showComment=1672760209887#c1463938625684108262) on a blog post by Ken Shirriff about the 8086's prefetch queue. Ken wrote that he doesn't know if it's the 486, Pentium or P6 generation when self-modifying code became effective instantaneously. I posted that we know for a fact that on the 486, there's still a minimum distance between code-modifying instructions, but nobody (in this thread, at least) looked at Pentiums.

So, to be clear, this has nothing to do with ao486. It's trying to mimic some generic 486-class CPU, and it's doing this just fine. It's an unrelated question.
I'm going to prepare a bootable disk image with the test program for @dwillmore (and maybe others who are interested in this stuff) and then close this issue, because the problem why it was opened years ago is indeed resolved.

**Comment by kajott** [comment]
Final revision of the test application: [smctest3.zip](https://github.com/MiSTer-devel/ao486_MiSTer/files/10421087/smctest3.zip), now including a 3.5" 1.44M FreeDOS-based boot disk image. If anyone has interesting results to share when running this on real hardware, please do so [in private](mailto:keyj@emphy.de) or as a comment below the relevant [post](http://www.righto.com/2023/01/inside-8086-processors-instruction.html) on Ken Shirriff's website, as this issue will now be closed.
